File: current_session/c++/1793.cpp
#include <iostream>
#include <vector>

using namespace std;

class Solution {
public:
  int maximumScore(vector<int>& nums, int k) {
    long maxScore = long(nums[k]);
    long minScore = long(nums[k]);
    int l = k, r = k;
    while (l>0 || r<nums.size()-1) {
      if (l==0 || (nums[l-1] < nums[r+1])) {
        minScore = min(minScore, long(nums[++r]));
      }
      else {
        minScore = min(minScore, long(nums[--l])) ;
      }
      maxScore = max(maxScore, minScore*(r-l+1));
    }
    return maxScore;
  }
};

File: current_session/python/1793.py
"""
1793
"""
from typing import List
class Solution:
    def maximumScore(self, nums: List[int], k: int) -> int:
        # Naive approach, try all L and R (including nums[k]), and use a heap to store min(subArray)
        # time complexity O(n^2 logn)
        # think about O(n) approach
        # since the 'good array' must include nums[k]
        # we can begin from nums[k]
        # increase l or r per iteration
        ans, minScore, l, r = nums[k], nums[k], k, k
        while l > 0 or r < len(nums)-1:
            if (l == 0):
                r += 1
            elif (r == len(nums)-1):
                l -= 1
            elif (nums[l-1] < nums[r+1]):
                r += 1
            else:
                l -= 1
            minScore = min(minScore, nums[l], nums[r])
            ans = max(ans, minScore*(r-l+1))
        return ans
